src/js/game/LevelMVC/LevelModel.js

```diff
@@ -581,17 +581,13 @@ module.exports = class LevelModel {
       // Prevent walking into water/lava in levels where the player is
       // controlled by arrow keys. In levels where the player is controlled by
       // blocks, let them drown.
-      if (this.groundPlane.getBlockAt(position).blockType === "water") {
+      const blockTypeAtPosition = this.groundPlane.getBlockAt(position).blockType;
+      const frontEntity = this.getEntityAt(position);
+      if (['water', 'lava'].includes(blockTypeAtPosition)) {
         if (this.controller.getIsDirectPlayerControl()) {
-          result.push("water");
+          result.push(blockTypeAtPosition);
         } else {
-          return [true];
-        }
-      } else if (this.groundPlane.getBlockAt(position).blockType === "lava") {
-        if (this.controller.getIsDirectPlayerControl()) {
-          result.push("lava");
-        } else {
-          return [true];
+          return [frontEntity === undefined || frontEntity.canMoveThrough()];
```

Comment:
are we sure we want to end early here, and not with a `notEmpty` or `notWalkable` or something? It's not 100% clear to me that a value of just `[false]` is valid

Reply:
That's a good point - it'd be a smaller change if we _don't_ early-out when the new condition is not met.  I'll give that a shot.